File: 卡码/代码随想录/JS/二叉树/02-二叉树属性/LeetCode_0513.js
//513. 找树左下角的值
import TreeNode from "../TreeNode.js";

let maxDepth=-1,res=0;
const traversal=(root,depth)=>{
  if(!root.left&&!root.right){
    if(depth>maxDepth){
      maxDepth=depth;
      res=root.val;
    }
    return;
  }
  root.left&&traversal(root.left,depth+1);
  root.right&&traversal(root.right,depth+1);
  return;
};
const findBottomLeftValue = function(root) {
  maxDepth=-1,res=0;
  traversal(root,0);
  return res;
};

//const findBottomLeftValue = function(root) {
//  const q=new Array();
//  if(root) q.push(root);
//  let res=0;
//  while(q.length>0){
//    const size=q.length;
//    for(let i=0;i<size;i++){
//      const node=q.shift();
//      if(i==0) res=node.val;
//      node.left&&q.push(node.left);
//      node.right&&q.push(node.right);
//    }
//  }
//  return res;
//};

File: 卡码/代码随想录/Go/二叉树/02-二叉树属性/LeetCode_0513.go
//513. 找树左下角的值
package main

import (
	. "binaryTree/types"
)

var maxDepth,res=-1,0
func traversal(root *TreeNode,depth int){
	if root.Left==nil&&root.Right==nil {
		if depth>maxDepth {
			maxDepth=depth
			res=root.Val
		}
		return
	}
	if root.Left!=nil {
		traversal(root.Left,depth+1)
	}
	if root.Right!=nil {
		traversal(root.Right,depth+1)
	}
	return
}
func findBottomLeftValue(root *TreeNode) int {
	maxDepth,res=-1,0
	traversal(root,0)
	return res
}

// func findBottomLeftValue(root *TreeNode) int {
// 	q:=make([]*TreeNode,0)
// 	if root!=nil {
// 		q=append(q,root)
// 	}
// 	res:=0
// 	for len(q)>0 {
// 		size:=len(q)
// 		for i:=0;i<size;i++ {
// 			node:=q[0]
// 			q=q[1:]
// 			if i==0 {
// 				res=node.Val
// 			}
// 			if node.Left!=nil {
// 				q=append(q,node.Left)
// 			}
// 			if node.Right!=nil {
// 				q=append(q,node.Right)
// 			}
// 		}
// 	}
// 	return res
// }
